<a href="https://colab.research.google.com/github/Dawaman43/3d-weather-app/blob/main/Python_datascience.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!git https://github.com/Dawaman43/my-colab-project.git

git: 'https://github.com/Dawaman43/my-colab-project.git' is not a git command. See 'git --help'.


In [ ]:
!git clone https://github.com/Dawaman43/my-colab-project.git

Cloning into 'my-colab-project'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (3/3), done.


# 1. what is the difference list , a dictionary and a numpy array

#list
List is ordered collection of items
List can hold any data type
List can be indexed

#dictionary
Dictionary is unordered collection
Stores key-value

#numpy array
Load from numpay library
Stores same data type

#2. we can use list comprehension to loop through given list and get desirable outcome

In [ ]:
def sq():
    sqr = [x**2 for x in [10, 15, 20, 25, 30] if x % 2 == 0]
    return sqr

#3. Code output

In [ ]:
[1, 2, 3, 4]

#4. What does .shape and .describe() tell you about a pandas DataFrame? Show using an example.

.shape() - returns tuple of dataframe like the count of dataframe in rows and columns

.describe() - returns summary of dataframe

In [ ]:
import pandas as pd


data = {
    "A": [1, 2, 3, 4],
    "B": [10, 20, 30, 40],
    "C": ["x", "y", "z", "w"]
}
df = pd.DataFrame(data)


print("Shape:", df.shape)


print("\nDescribe:\n", df.describe())

Shape: (4, 3)

Describe:
         A     B
count  4.0   4.0
mean   2.5  25.0
std    1.3  12.9
min    1.0  10.0
25%    1.8  17.5
50%    2.5  25.0
75%    3.3  32.5
max    4.0  40.0

#5. Load 5 rows from given csv

I see some difference the first row of all elements quoted

In [ ]:
import pandas as pd
df = pd.read_csv('airtel.csv')
print(df.head(5))

#6. Code problem from question 5

In [ ]:
import pandas as pd

df = pd.read_csv('airtravel.csv')


df['Total'] = df.iloc[:, 1:].sum(axis=1)

max_month = df.loc[df['Total'].idxmax()]

print("Month with highest total passengers:")
print(max_month)

In [ ]:
import pandas as pd

df = pd.read_csv('airtravel.csv')

min_month_1958 = df.loc[df['1958'].idxmin()]

print("Month with the lowest passengers in 1958:")
print(min_month_1958)

#7. What does .groupby() function do in pandas? Show an example using dummy data

.groupby() - groups rows of a dataframe based on  one or more columns and to make easy aggregate operation like sum, mean

In [ ]:
import pandas as pd


data = {
    "Team": ["A", "A", "B", "B", "C"],
    "Score": [10, 20, 15, 25, 30]
}
df = pd.DataFrame(data)


grouped = df.groupby('Team')['Score'].sum()
print(grouped)

Output :
Team
A    30
B    40
C    30
Name: Score, dtype: int64

# 8. Count missing values from each coumn in titanic dataset

In [7]:
import seaborn as sns
import pandas as pd

df = sns.load_dataset('titanic')

missing_values = df.isnull().sum()
print("Missing values in each column:\n", missing_values)

Missing values in each column:
 survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


Missing values in each column:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

#9 visualise df["Age"].hist()

1. Since .hist() method is called it generated histogram of Age dataframe so for example if we have , [2,3,7,11] it may generate [0-5], [5-10] and [10-15]

2. Sorry

#10 What does it mean if a column has a high standard deviation?

Standard deviation (std. dev.) Is measurement of the spread of the dataset relative to the mean. So high standard deviation means the data is spread out the mean or the data has a lot of variability as the value spread far from the mean.
__example__ - List: [1, 5, 10, 20, 64]

Mean = (1 + 5 + 10 + 20 + 64) / 5 = 100 / 5 = 20

Deviations from mean: [-19, -15, -10, 0, 44]Variance = (-19² + -15² + -10² + 0² + 44²) / 5 = (361 + 225 + 100 + 0 + 1936) / 5 = 2622 / 5 = 524.4

Std. Dev. = √524.4 ≈ 22.9 (high, as values are far from the mean).

#11 What are some real-world problems where missing data might affect results? Give 2 examples.

1. __medical research(trial)__ - if data doesn't collected correctly from patients or persons it may affect the medical process even it may cause death

2. __customer servey__ - collecting information about the company from the customer may go wrong if it is not collected correctly , it may even cause bankrupt of the company